File: practice/src/main/java/com/decade/practice/usecases/ChatEventStore.java
package com.decade.practice.usecases;

import com.decade.practice.data.repositories.ChatRepository;
import com.decade.practice.models.domain.entity.Chat;
import com.decade.practice.models.domain.entity.ChatEvent;
import com.decade.practice.models.domain.entity.MessageEvent;
import com.decade.practice.models.domain.entity.User;
import com.decade.practice.utils.ChatUtils;
import org.hibernate.exception.ConstraintViolationException;
import org.springframework.context.annotation.Primary;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Isolation;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.Collection;
import java.util.NoSuchElementException;

@Transactional(isolation = Isolation.READ_COMMITTED)
@Component
@Primary
public class ChatEventStore implements EventStore {

        private final UserEventStore userEventStore;
        private final ChatRepository chatRepo;
        private final ChatOperations chatOperations;

        public ChatEventStore(UserEventStore userEventStore, ChatRepository chatRepo, ChatOperations chatOperations) {
                this.userEventStore = userEventStore;
                this.chatRepo = chatRepo;
                this.chatOperations = chatOperations;
        }

        @Override
        public Collection<ChatEvent> save(ChatEvent event) throws NoSuchElementException, ConstraintViolationException {
                // Ensure chat exists before locking/updating
                Chat chat = chatOperations.getOrCreateChat(event.getChatIdentifier());
                if (event instanceof MessageEvent) {
                        chat.setMessageCount(chat.getMessageCount() + 1);
                }
                event.setChat(chat);

                User me = ChatUtils.inspectOwner(chat, event.getSender());
                User you = ChatUtils.inspectPartner(chat, me);

                ChatEvent mine = event.copy();
                mine.setLocalId(event.getLocalId());

                ChatEvent yours = event.copy();

                mine.setOwner(me);
                yours.setOwner(you);

                Collection<ChatEvent> result = new ArrayList<>();
                result.addAll(userEventStore.save(mine));
                result.addAll(userEventStore.save(yours));
                return result;
        }
}

File: practice/src/main/java/com/decade/practice/database/transaction/ChatEventStore.kt
package com.decade.practice.database.transaction

import com.decade.practice.core.ChatOperations
import com.decade.practice.core.EventStore
import com.decade.practice.model.entity.ChatEvent
import com.decade.practice.model.entity.isMessage
import com.decade.practice.util.inspectOwner
import com.decade.practice.util.inspectPartner
import jakarta.persistence.EntityManager
import jakarta.persistence.LockModeType
import jakarta.persistence.PersistenceContext
import org.hibernate.exception.ConstraintViolationException
import org.springframework.context.annotation.Primary
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Isolation
import org.springframework.transaction.annotation.Transactional

@Component
@Primary
class ChatEventStore(
      private val eventStore: UserEventStore,
      private val chatOperations: ChatOperations,
) : EventStore {

      @PersistenceContext
      lateinit var em: EntityManager

      @Transactional(isolation = Isolation.READ_COMMITTED)
      @Throws(
            NoSuchElementException::class,
            ConstraintViolationException::class
      )
      override fun save(event: ChatEvent): Collection<ChatEvent> {
            val chat = chatOperations.getOrCreateChat(event.chatIdentifier)
            em.lock(chat, LockModeType.PESSIMISTIC_WRITE)
            if (event.isMessage()) {
                  chat.messageCount++
            }
            event.chat = chat

            val me = chat.inspectOwner(event.sender)
            val you = chat.inspectPartner(me)

            val mine = event.copy().apply {
                  localId = event.localId
            }
            val yours = event.copy()

            mine.owner = me
            yours.owner = you

            return eventStore.save(mine) + eventStore.save(yours)
      }

}
